### PR TITLE
oracle-function: implement trunc and trunc_datetime

### PR DESCRIPTION
**Pull request by fishiu**
### What problem does this PR solve?

Issue Number: ref #57815

Problem Summary:

### What changed and how does it work?

In trunc function, we handle differently based on type of the first arguments:

If the first argument is number, i.e. `trunc(num1[, num2])`, it will be a number truncate and we reuse the builtin TRUNCATE function.

If the first argument is datetime, i.e. `trunc(datetime[, fmt])`, it will return a datetime truncated to the unit described in fmt. See details in the following table.

Here's the truncation rules for the TRUNC function on datetime:
| Format | Aliases | Preserved Parts | Sample Input | Sample Output |
|--------|---------|----------------|--------------|---------------|
| YEAR | YY, YYYY | year | 2024-11-27 22:03:45 | 2024-01-01 00:00:00 |
| MONTH | MM | year, month | 2024-11-27 22:03:45 | 2024-11-01 00:00:00 |
| DAY | DD | year, month, day | 2024-11-27 22:03:45 | 2024-11-27 00:00:00 |
| HOUR | HH | year, month, day, hour | 2024-11-27 22:03:45 | 2024-11-27 22:00:00 |
| MINUTE | MI | year, month, day, hour, minute | 2024-11-27 22:03:45 | 2024-11-27 22:03:00 |

Notes:
- Fmt is case-insensitive
- We truncate to day by default
- Truncation to second is not supported
- Invalid format will return an error

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/expression/builtin_math.go [hunk 1]
```diff
@@ -2078,3 +2079,44 @@ func (b *builtinTruncateUintSig) evalInt(ctx EvalContext, row chunk.Row) (int64,
 	shift := uint64(math.Pow10(int(-d)))
 	return int64(uintx / shift * shift), false, nil
 }
+
+type truncFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *truncFunctionClass) getFunction(ctx BuildContext, args []Expression) (builtinFunc, error) {
```
Comment (joccau) [inline pkg/expression/builtin_math.go:2087]:
```
maybe check it as blow code ?

arg0Tp := args[0].GetType(ctx.GetEvalCtx()).GetType();
evalTp := arg0Tp.EvalType()

switch evalTp {
    case types.ETDatetime:
        // deal parameter with Datetime
        // need to implement a builtin，like builtinStrToDateDatetimeSig
       // code
    case types.ETInt, types.ETREAL, type.ETDeciaml:
        // deal parameter with numeric data.
        fc := truncateFunctionClass{baseFunctionClass{ast.Truncate, 2, 2}}
        return fc.getFunction(ctx, args).
    default:
		return nil, perrors.Errorf("invalid argument type %v", arg0Tp)
} 
```

**Review by joccau — APPROVED** [review]
Look pretty good~

**Review by joechenrh — APPROVED** [review]

**Review by Benjamin2037 — APPROVED** [review]
LGTM

**Pull request merged**